text/0000-disjoint-groups/0000-disjoint-groups.md

```diff
@@ -0,0 +1,149 @@
+
+### Banning nodes from the network
+
+The routing tables of all nodes in a group will be identical. This allows the network to ban misbehaving nodes: If a node's group decides to ban the node from the network, it knows exactly whom this node is connected to, and who needs to be notified to make the whole network disconnect from that it.
+
+### Weigthed votes and node evaluation
+
+If group `A` is connected to group `B`, every node in `B` will know every member of `A`, so group `B` can easily reach a consensus regarding the correctness of the individual `A`-members' behaviour. This allows `B` to evaluate the nodes in `A` and give feedback about them - another requirement for banning misbehaving nodes. But it could be taken one step further and used to assign grades, and make a nodes' vote weight for group consensus depend on it: Making voting rights depend on a proof of work is essential for the security of the network.
+
+### Data chains
+
+The current group definition would considerably complicate [data chains](https://github.com/dirvine/data_chain): With disjoint groups, all members of a group can sign the same link block - consising of all the group members -, and are responsible for exactly the same set of data chunks, so they can sign the same data blocks.
+
+
+## Detailed design
+
+The definition of a close group is modified so that at each point in time, the name space is partitioned into disjoint groups. Each group is defined by a _name prefix_, i.e. a sequence of between 0 and 256 bits, similar to how IP subnets correspond to IP address prefixes. The group `(p)` consists of all nodes whose name begins with the prefix `p`, and is responsible for all data items whose name begins with `p`. The group thus manages the part of the name space given by the interval `[p00...00, p11...11]`.
+
+When the network is bootstrapped, there is only one group, with the empty prefix `()`, responsible for the whole name space.
+
+* Whenever `(p0)` and `(p1)` satisfy certain group requirements (see below), a group `(p)` splits into `(p0)` and `(p1)`.
+* Whenever a group `(p0)` or `(p1)` ceases to satisfy the group requirements, it merges with all its sister groups into the group `(p)` again: All groups that would be a subset of `(p)` are merged back into `(p)`.
+
+These are the only two operations allowed on groups, so it is guaranteed that every address in the network belongs to exactly one group, and that group satisfies the requirements.
+
+The invariant that needs to be satisfied by the routing table is modified accordingly:
+
+1. A node must have its complete group `(p)` in its routing table.
+2. It must be connected to every group `(q)`, for which `p` and `q` differ in exactly one bit.
```

Comment:
Should this be "fully connected" or connected to each node, or are we considering and advanced mechanism perhaps to allow proxy connections or similar?